### PR TITLE
tinygo #2979 update tinygo build constraints

### PR DESCRIPTION
**Pull request by archie2x**
A lot of the tinygo build constraints were out of date with tinygo dev branch. Many commands are now compile/linkable. YMMV running them.

1. update tinygoize s.t. it will insert a `go:build` line if one is not already present
2. `cmds/*/*` with up-to-date tinygo build constraints 
3. update the tinygo build constraint as`!tinygo || tinygo.enable` s.t. you can build anything with compilation disabled by supplying the `tinygo.enable` tag without having to think through all the other build constraints for a given command.

```sh
# build a cmd currently tinygo-disabled
tinygo build -tags tinygo.enable
```

### CODE REVIEW

**Comment by archie2x** [comment]
> this seems fine to me?

I think it's out of date with current tinygo-dev but I find it really useful to use `-tags tinygo.enable` rather than working through each package's build tags logic individually.

I did some more modification to tinygoize to auto-update the build tags and progress README but using it hereafter will produce very small changesets.

File: tools/tinygoize/main.go [hunk 1]
```diff
@@ -2,81 +2,315 @@
...
+	return
+}
+
+func writeMarkdown(file *os.File, pathMD *string, tinygo *string, status BuildStatus) (err error) {
+
+	fmt.Fprintf(file, "---\n\n")
+	fmt.Fprintf(file, "DO NOT EDIT.\n\n")
+	fmt.Fprintf(file, "Generated via `go run tools/tinygoize/main.go`\n\n")
+	fmt.Fprintf(file, "%v\n\n", tinygoVersion(tinygo))
+	fmt.Fprintf(file, "---\n\n")
```
Comment (leongross) [inline tools/tinygoize/main.go:217]: Is there a reason you separate this over multiple `Fprintf`?
Why not define sth like [this](https://github.com/u-root/u-root/blob/1a80242f1fee21955d85c1e892c78e20a31ddf5f/cmds/exp/bzimage/bzimage.go#L40) with a format specifier and then do that? imo this would make it more readable

File: tools/tinygoize/main.go [hunk 2]
```diff
@@ -2,81 +2,315 @@
...
+## Commands Build Status
+`)
+
+	linkText := func(dir string) string {
+		relPath, _ := filepath.Rel(filepath.Dir(*pathMD), dir)
+		return fmt.Sprintf("[%v](%v)", dir, relPath)
+	}
+
+	processSet := func(header string, dirs []string) {
+
```
Comment (leongross) [inline tools/tinygoize/main.go:247]: nit
```suggestion

```

File: tools/tinygoize/main.go [hunk 3]
```diff
@@ -2,81 +2,315 @@
...
+The necessary additions to tinygo will be tracked in
+[#2979](https://github.com/u-root/u-root/issues/2979).
+
+---
+
+## Commands Build Status
+`)
+
+	linkText := func(dir string) string {
+		relPath, _ := filepath.Rel(filepath.Dir(*pathMD), dir)
```
Comment (leongross) [inline tools/tinygoize/main.go:242]: Why ignore the error here?

File: tools/tinygoize/main.go [hunk 4]
```diff
@@ -2,81 +2,315 @@
...
+	if nil != err {
+		log.Fatalf("version: %v", err)
+	}
+	return strings.TrimSpace(string(out))
+}
+
+// check (via `go build -n`) if a given directory would have been skipped
+// due to build constraints (e.g. cmds/core/bind only builds for plan9)
+func isExcluded(dir string) bool {
+	// to lazy to dynamically pull tags from `tinygo info`
```
Comment (leongross) [inline tools/tinygoize/main.go:69]:
```suggestion
	// too lazy to dynamically pull tags from `tinygo info`
```

File: tools/tinygoize/main.go [hunk 5]
```diff
@@ -2,81 +2,315 @@
...
-// For each directory, it will try a tinygobuild with
-// CGO_ENABLED=0, GOARCH=amd64, and GOOS=linux.
-// If the tinygo build fails, it will use ast package
-// to rewrite //go:build lines as follows:
-// the line starts as //go:build expr
-// it is rewritten to //go:build !tinygo && (expr)
-// When the file is written, the expression seems
-// to be simplified.
+// usage: invoke this with a list of directories. For each directory, it will
+// run `CGO_ENABLED=0 GOARCH=amd64 GOOS=linux tinygo build -tags tinygo.enable`
```
Comment (leongross) [inline tools/tinygoize/main.go:6]: Same with `CGO_ENABLED=0`

File: tools/tinygoize/main.go [hunk 6]
```diff
@@ -2,81 +2,315 @@
...
+
+// Track set of passing, failing, and excluded commands
+type BuildStatus struct {
+	passing  []string
+	failing  []string
+	excluded []string
+}
+
+// return trimmed output of "tinygo version"
+func tinygoVersion(tinygo *string) string {
```
Comment (leongross) [inline tools/tinygoize/main.go:58]: I don't think calling `Fatalf` is a clean way to exit here, since it calls `os.Exit(1)` under the hood. Reconsider returning an error and handle this the way up. I see that you did not introduce this but it was there before, but when we do a rework we might as well make it clean, right?

```suggestion
func tinygoVersion(tinygo *string) (string, error) {
```

**Review by leongross — CHANGES_REQUESTED** [review]

File: tools/tinygoize/main.go (hunk 1 above)
Comment (archie2x) [reply tools/tinygoize/main.go:217]: I did one line down from this. Here I didn't because of the markdown backticks in :

Generated via `go run tools/tinygoize/main.go`\n\n

The backticks conflict with the golang string-literal delimiter and (apparently!) no way to escape them.

File: tools/tinygoize/main.go [hunk 7]
```diff
@@ -2,81 +2,315 @@
...
+// "tinygo build" in directory 'dir'
+func build(tinygo *string, dir string) (err error) {
+	tags := []string{"tinygo.enable"}
+	if addTags := buildTags(dir); addTags != "" {
+		tags = append(tags, addTags)
+	}
+	c := exec.Command(*tinygo, "build", "-tags", strings.Join(tags, ","))
+	c.Dir = dir
+	c.Stdout, c.Stderr = os.Stdout, os.Stderr
+	c.Env = append(os.Environ(), "GOOS=linux", "CGO_ENABLED=0", "GOARCH=amd64")
```
[the remark this answers is not in the document]
Comment (archie2x) [reply tools/tinygoize/main.go:99]: I just removed it. Let it use the default.

File: tools/tinygoize/main.go [hunk 8]
```diff
@@ -2,81 +2,315 @@
...
+	return c.Run()
+}
+
+// "tinygo build" in each of directories 'dirs'
+func buildDirs(tinygo *string, dirs []string) (status BuildStatus, err error) {
```
[the remark this answers is not in the document]
Comment (archie2x) [reply tools/tinygoize/main.go:104]: Introduced better error handling.

File: tools/tinygoize/main.go [hunk 9]
```diff
@@ -2,81 +2,315 @@
...
+
+	for _, dir := range dirs {
+		log.Printf("Building %s\n", dir)
+		if berr := build(tinygo, dir); berr != nil {
```
[the remark this answers is not in the document]
Comment (archie2x) [reply tools/tinygoize/main.go]: Introduced better error handling.

File: tools/tinygoize/main.go (hunk 3 above)
Comment (archie2x) [reply tools/tinygoize/main.go:256]: Added comment.

File: tools/tinygoize/main.go (hunk 4 above)
Comment (archie2x) [reply tools/tinygoize/main.go]: Done

File: tools/tinygoize/main.go (hunk 6 above)
Comment (archie2x) [reply tools/tinygoize/main.go]: Done.

File: tools/tinygoize/main.go (hunk 5 above)
Comment (archie2x) [reply tools/tinygoize/main.go]: Removed.

**Review by leongross — APPROVED** [review]
I saw some minor formatting-nits were I was not sure if this could be done a bit more nicely, maybe run `gofumpt`. Apart from that lgtm.

**Comment by MDr164** [comment]
> EDIT: I saw the CI was failing due to some docker pulling issues, please re-run it so we can get the CI green :)

I retriggered it and now everything except Codecov passes. The reason for codecov to have such a big negative diff might be due to this branch not being in sync with main though.

**Review by leongross — APPROVED** [review]
lgtm

File: cmds/core/insmod/tinygo_tag_noasm_amd64.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017-2018 the u-root Authors. All rights reserved
```
Comment (rminnich) [inline cmds/core/insmod/tinygo_tag_noasm_amd64.go:1]: just checking, is there a test for this? I do have a test, badbuild, which is for things that should break a build, if you want to test it there.

**Review by rminnich — APPROVED** [review]

**Review by rminnich — APPROVED** [review]

File: cmds/core/insmod/tinygo_tag_noasm_amd64.go (hunk 10 above)
Comment (archie2x) [reply cmds/core/insmod/tinygo_tag_noasm_amd64.go:1]: There isn't. I was wondering however if we should add tests to compile everything that should be tinygo compilable to prevent regression there? Or aren't we there yet? Given the rapid progress of things it seems like it might be a good time for it.
Anyway, right now, I don't think there are any tests that use tinygo. I noticed that tinygo -dev builds are available now via nightlies.

Comment (archie2x) [reply cmds/core/insmod/tinygo_tag_noasm_amd64.go:1]: @leongross To build my tinygo busybox, I
(1) used makebb to build just {init,shutdown,gosh} as "go-bb" and put it in a pre-symlinked directory
(2) modified mkuimage so that it would honor the symlinks above
(3) built u-root with u-root/gobusybox#121 passing appropriate --init and --sh options (to point to go-bb)

File: tools/tinygoize/main.go [hunk 11]
```diff
@@ -2,81 +2,340 @@
...
+			}
+		}
+	}
+	return
+}
 
+// Modifies, adds, or removes //go:build line as appropriate to include / remove
+// '!tinygo || tinygo.enable' for all .go files in dir depending on whether it
+// 'builds' or not as previously tested.
+func fixupConstraints(dir string, builds bool) (err error) {
```
Comment (leongross) [inline tools/tinygoize/main.go:136]: The functionality this function provides adds a lot of complex functionality. Especially since it works on the AST of the go programs and does literal modifications and checks a lot of edge cases we should add some basic test cases to ensure the modifications are done correctly. Since this PR has been in review for quite a while and the except my suggestions look good to me, we could also bring in the tests in a follow-up PR.

**Review by leongross — CHANGES_REQUESTED** [review]

**Comment by leongross** [comment]
@archie2x sorry for being late with the review. I added a new batch of suggestions (also many unimportant nits). When this is fixed we can get this merged.

File: tools/tinygoize/main.go [hunk 12]
```diff
@@ -2,81 +2,340 @@
...
+					continue nextFile
+				}
+
+				if builds {
+					re := regexp.MustCompile(`\(?\s*!tinygo\s+\|\|\s+tinygo.enable\s*\)?(\s+\&\&)?`)
+					c.Text = re.ReplaceAllString(c.Text, "")
+					log.Printf("Stripping build constraint %v\n", file)
+
+					// handle potentially now-empty build constraint
+					re = regexp.MustCompile(`^\s*//go:build\s*$`)
```
[the remark this answers is not in the document]
Comment (archie2x) [reply tools/tinygoize/main.go:183]: back tick strings?

File: tools/tinygoize/main.go (hunk 11 above)
Comment (archie2x) [reply tools/tinygoize/main.go:136]: Yeah, might get to some units on the AST checks.  Shouldn't run this thing on anything but a clean tree and shouldn't commit anything it modifies without visual inspection. In that vein, not sure of the value of units. Part of me finds this an overly complicated replacement for a perl one-liner.

**Comment by archie2x** [comment]
Please see #3107 where I pulled out just the `//go:build` constraint and README changes.
